lightning/src/routing/router.rs

```diff
@@ -876,40 +877,115 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+					let hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+						Some(htlc_minimum_msat) => htlc_minimum_msat,
+						None => 0
+					};
+
+					let directional_info = DummyDirectionalChannelInfo {
+						cltv_expiry_delta: hop.cltv_expiry_delta as u32,
+						htlc_minimum_msat: hop_htlc_minimum_msat,
+						htlc_maximum_msat: hop.htlc_maximum_msat,
+						fees: hop.fees,
+					};
+
+					prev_hop = match idx {
+						0 => &payee,
+						_ => match prev_route_it.next() {
+							Some(x) => &x.src_node_id,
+							None => &payee,
+						}
+					};
+					let reqd_channel_cap = Some(match idx {
+						0 => (final_value_msat + 999),
+						_ => match aggregate_next_hops_fee_msat.checked_add(999) {
+							Some(cap) => cap,
+							None => u64::max_value()
+						},
+					} / 1000); // converting from msat
+
+
+					// We assume that the recipient only included route hints for routes which had
+					// sufficient value to route `final_value_msat`. Note that in the case of "0-value"
+					// invoices where the invoice does not specify value this may not be the case, but
+					// better to include the hints than not.
+					if add_entry!(hop.short_channel_id, hop.src_node_id, prev_hop, directional_info, reqd_channel_cap, &empty_channel_features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat) {
+
+						// If the index is 0, then its the last hop in the RoutHint. The
+						// aggregate_next_hops_fee_msat for this hop was 0 but from here on out
+						// it needs to be increased beyond the value to be routed to account
+						// for the fee charged at hops and value to send from payer to payee.
+						if idx == 0 {
+							aggregate_next_hops_fee_msat = match aggregate_next_hops_fee_msat.checked_add(final_value_msat) {
+							Some(value) => value,
+							None => u64::max_value()
+							};
+						}
+					} else {
+						// If this hop was not used then there is no use checking the preceeding hops
+						// in the RouteHint. We can break by just searching for a direct channel between
+						// last checked hop and first_hop_targets
+						if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&prev_hop) {
+							add_entry!(first_hop, *our_node_id , prev_hop, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, aggregate_next_hops_fee_msat, path_value_msat, aggregate_next_hops_path_htlc_minimum_msat);
+						}
+						break;
 					}
+
+					if idx == route.0.len() - 1 {
+						// The last hop in this iterator is the first hop in
+						// overall RouteHint.
+						// If this hop connects to a node with which we have a direct channel,
+						// ignore the network graph and, if the last hop was added, add our
+						// direct channel to the candidate set.
+						//
+						// Note that we *must* check if the last hop was added as `add_entry`
+						// always assumes that the third argument is a node to which we have a
+						// path.\
```

Comment:
nit: spare `\` at EOL here.

Reply:
Regarding moving pre-```break``` ```add_entry!```  and out doing it always, that would mean checking for a direct channel between all node hops in ```RouteHint``` and I would also need to use another variable (eg, ```hop_used```) just to track if the hop was used or not and then break accordingly out of the loop after the ```add_entry!``` with the previous ```RouteHintHop```.

Reply:
Hmm, yea, it does feel clearer if you do that, I think. If we are going to look up first-hop connections in some cases, we might as well look it up in all cases.

Reply:
From commit [0c0985](https://github.com/rust-bitcoin/rust-lightning/pull/1040/commits/0c09852311b3b7c42d1210aa69aa3f827a91655d) all hops are checked for a direct channel.